lib/bolt/plugin_cache.rb

```diff
@@ -0,0 +1,70 @@
+# frozen_string_literal: true
+
+require 'fileutils'
+require 'bolt/error'
+require 'bolt/util'
+
+module Bolt
+  class PluginCache
+    attr_reader :reference, :cache_file, :default_config, :id
+
+    def initialize(reference, cache_file, default_config)
+      @reference = reference
+      @cache_file = cache_file
+      @default_config = default_config
+    end
+
+    def read_and_clean_cache
+      return if ttl == 0
+      validate
+
+      # Luckily we don't need to use a serious hash algorithm
+      require 'digest/bubblebabble'
+      r = reference.select { |k, _| k == '_cache' }.sort.to_s
+      @id = Digest::SHA2.bubblebabble(r)[0..20]
+
+      unmodified = true
+      # First remove any cache entries past their ttl
+      # This prevents removing plugins from leaving orphaned cache entries
+      cache.delete_if do |_, entry|
+        expired = Time.now - Time.parse(entry['mtime']) >= entry['ttl']
+        unmodified = false if expired
+        expired
+      end
+      File.write(cache_file, cache.to_json) unless cache.empty? || unmodified
+
+      cache.dig(id, 'result')
+    end
+
+    private def cache
+      @cache ||= Bolt::Util.read_optional_json_file(@cache_file, 'cache')
+    end
+
+    def write_cache(result)
+      cache.merge!({ id => { 'result' => result,
+                             'mtime' => Time.now,
+                             'ttl' => ttl } })
+      FileUtils.touch(cache_file)
+      File.write(cache_file, cache.to_json)
+    end
+
+    def validate
+      # The default cache `plugin-cache` will be validated by the config
+      # validator
+      return if reference['_cache'].nil?
+      r = reference['_cache']
+      unless r.is_a?(Hash)
+        raise Bolt::ValidationError,
+              "_cache must be a Hash, received #{r.class}: #{r.inspect}"
+      end
+
+      unless r.key?('ttl')
+        raise Bolt::ValidationError, "_cache must set 'ttl' key."
+      end
```

Comment:
We'll want to validate that `ttl` is >= 0 here, since the config validator won't validate references.